pkg/controller/apimanagerbackup/apimanagerbackup_logic_reconciler.go

```diff
@@ -0,0 +1,398 @@
+		if err != nil {
+			return nil, err
+		}
+		apiManagerBackup := backup.NewAPIManagerBackup(options)
+
+		res.apiManagerBackup = apiManagerBackup
+	}
+
+	return res, nil
+}
+
+func (r *APIManagerBackupLogicReconciler) Logger() logr.Logger {
+	return r.logger
+}
+
+func (r *APIManagerBackupLogicReconciler) Reconcile() (reconcile.Result, error) {
+	if r.cr.BackupCompleted() {
+		r.Logger().Info("Backup completed. End of reconciliation")
+		return reconcile.Result{}, nil
+	}
+
+	result, err := r.reconcileAPIManagerSourceStatusField()
+	if result.Requeue || err != nil {
+		return result, err
+	}
+
+	result, err = r.reconcileStartTimeField()
+	if result.Requeue || err != nil {
+		return result, err
+	}
+
+	result, err = r.reconcileBackupInS3Destination()
+	if result.Requeue || err != nil {
+		return result, err
+	}
+
+	result, err = r.reconcileBackupInPVCDestination()
+	if result.Requeue || err != nil {
+		return result, err
+	}
+
+	result, err = r.reconcileBackupCompletion()
+	if result.Requeue || err != nil {
+		return result, err
+	}
+
+	return result, err
+}
+
+func (r *APIManagerBackupLogicReconciler) reconcileBackupInS3Destination() (reconcile.Result, error) {
+	// TODO implement
+	return reconcile.Result{}, nil
+}
+
+func (r *APIManagerBackupLogicReconciler) reconcileBackupInPVCDestination() (reconcile.Result, error) {
+	var res reconcile.Result
+	var err error
+
+	err = r.reconcileBackupDestinationPVC()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	res, err = r.reconcileBackupDestinationPVCStatus()
+	if res.Requeue || err != nil {
+		return res, err
+	}
+
+	res, err = r.reconcileBackupSecretsAndConfigMapsToPVCJob()
+	if res.Requeue || err != nil {
+		return res, err
+	}
+
+	res, err = r.reconcileAPIManagerCustomResourceBackupToPVCJob()
+	if res.Requeue || err != nil {
+		return res, err
+	}
+
+	res, err = r.reconcileBackupSystemFileStoragePVCToPVCJob()
+	if res.Requeue || err != nil {
+		return res, err
+	}
+
+	return res, err
+}
+
+func (r *APIManagerBackupLogicReconciler) reconcileBackupDestinationPVC() error {
+	desired := r.apiManagerBackup.BackupDestinationPVC()
+	if desired == nil {
+		return nil
+	}
+
+	// TODO create mutator function for PVC ?
+	err := r.ReconcileResource(&v1.PersistentVolumeClaim{}, desired, reconcilers.CreateOnlyMutator)
+
+	return err
+}
+
+func (r *APIManagerBackupLogicReconciler) reconcileBackupSecretsAndConfigMapsToPVCJob() (reconcile.Result, error) {
+	desired := r.apiManagerBackup.BackupSecretsAndConfigMapsToPVCJob()
+	if desired == nil {
+		return reconcile.Result{}, nil
+	}
+
+	// TODO create Mutator function for Jobs??
+	existing := &batchv1.Job{}
+	// Check if this step has finished
+	if r.cr.SecretsAndConfigMapsBackupStepFinished() {
+		return reconcile.Result{}, nil
+	}
+
+	// Check if backup substep has completed
+	if !r.cr.SecretsAndConfigMapsBackupSubStepFinished() {
+		err := r.ReconcileResource(existing, desired, reconcilers.CreateOnlyMutator)
```

Comment:
I think it is quite ugly to use `existing`, although it works for now.

I would not have implemented with `ReconcileResource` method because Jobs do not need reconcilliation. 

I would have done

```
r.GetResource(NamespacedName(desired), existing)
if notFound(){
     r.CreateREsource(desired)
     // you can expect the job to not be finished instantly, so requeue
      return reconcile.Result{Requeue: true}, nil
}

// continue with the existing object checking for completion...
```

Reply:
Ok I'll have it in the back of my mind. I might change it. Let's leave it this way for now until other parts are discussed :+1: